[MRG] Fix PDB reader + add argument to toggle proximity bonding

mwojcikowski opened the pull request:
#### What does this implement/fix? Explain your changes.
~~Disclaimer - this is still "work-in-progress".~~
This aims to fix issues with parsing PDB files by RDKit, such as:

- [x] covalent bonding with metals
- [x] multivalent Hs (bonded to many heavy atoms via `ConnectTheDots`)
- [x] covalent bonds with HOH residues
- [x] add an option to disable automated bonding, if the molecule was once sanitized, but for some reasons gets bad bonds again, or has pregenerated bonding

#### Any other comments?
These are the most common issues that we found on PDBBind refined set.
Do you know of any other common issues I could add to that TODO?

~~I'm still struggling to make the `proximityBonding` parameter work as it should (disable `ConnectTheDots`). I think I've missed some place and it gets ignored, any ideas?~~
My conda went crazy, I think everything is OK with the parameter and it works.

mwojcikowski commented:
I've done some extra testing on PDBbind refined set v2016 (4057 protein-ligand complexes). Current stable 2017.09.1 sanitizes 3459 and this PR sanitizes 3776 successfully, which is quite good improvement given the simplicity of addition.

First, tests suggest that all other problems can be resolved by templates of PDB residues (will discuss this in other PR/Issue).

mwojcikowski commented:
This PR is ready for review. Here are some final stats on PDBbind database v2016. I think it's remarkable improvement, I was not expecting such improvement. Ping @greglandrum 

I still need to validate in depth those proteins that fail. The sanitization of a protein does not guarantee that the residues have reasonable bonding, but a fix to that issue is yet to come and is out of scope of this PR.

Version | Refined Set | Sanitized Pockets - Refined Set | Sanitized Proteins - Refined Set | General Set | Sanitized Pockets - General Set | Sanitized Proteins - General Set
---------- | --------------- | ------------------------------------------- | ------------------------------------------ | ----------------- | ------------------------------------------ | --------------------------------------------
PR | 4057 | 4031 | 3895 | 13283 | 13222 | 12833
2017.09.1 | 4057 | 3459 | 2087 | 13283 | 11302 | 6984
improvement | - | 572 | 1808 | - | 1920 | 5849

mwojcikowski commented:
I have found yet another easy fix - unbinding the HOH residues from protein. Here are updated results on PDBbind.

Version | Refined Set | Sanitized Pockets - Refined Set | Sanitized Proteins - Refined Set | General Set | Sanitized Pockets - General Set | Sanitized Proteins - General Set
---------- | --------------- | ------------------------------------------- | ------------------------------------------ | ----------------- | ------------------------------------------ | --------------------------------------------
PR | 4057 | 4051 | 3995 | 13283 | 13259 | 13047
2017.09.1 | 4057 | 3459 | 2087 | 13283 | 11302 | 6984
improvement | - | 592 | 1908 | - | 1957 | 6063

mwojcikowski commented:
It would be nice to merge #1280 as well

marta-sd commented:
I checked how these changes affect performance. I compared this version to the current release on 157 protein structures from the PDBbind v2016 core set (i.e. the structures I was able to load with the current release). The new version is 2% slower on average. I thinks this difference is insignificant when we think of the benefits.

greglandrum commented:
@mwojcikowski , @marta-sd : thanks for the effort here!

Now that these are ready to be reviewed, should I start with #1280 and then do this one or would you prefer that I start here?

greglandrum commented on Code/GraphMol/FileParsers/MolSupplier.h:
```diff
@@ -345,7 +346,7 @@ class PDBMolSupplier : public MolSupplier {
 
  protected:
-  bool df_sanitize, df_removeHs;
+  bool df_sanitize, df_removeHs, d_proximityBonding;
```
should be `df_proximityBonding` to be consistent.

greglandrum commented on Code/GraphMol/FileParsers/ProximityBonds.cpp:
```diff
@@ -32,6 +32,21 @@ struct ProximityEntry {
 static bool IsBonded(ProximityEntry *p, ProximityEntry *q, unsigned int flags) {
   if (flags & ctdIGNORE_H_H_CONTACTS && p->elem == 1 && q->elem == 1)
     return false;
+  // do not make bonds to metals
```
this would also use the function mentioned above.

mwojcikowski commented on Code/GraphMol/FileParsers/PDBParser.cpp:
```diff
@@ -296,6 +296,23 @@ static void PDBBondLine(RWMol *mol, const char *ptr, unsigned int len,
         fail = true;
       }
+      // do not make bonds to metals
+      unsigned int elem = amap[src]->getAtomicNum();
```
Yes, we can also include halogen anions as well. Although these would require further checkup (e.g. if the residue number is different).

mwojcikowski replied:
I will move this code out to a separate function, although both blocks serve different purpose. The one in `PDBBondLine()` avoids adding non-covalent bonds, which usually are present in PDB files (the official ones from wwPDB). The other one, in `IsBonded()`, avoidscreating bonds by the RDKit's `ConnectTheDots()`, which connects every pair of atoms which is closer than the sum of covalent radii. I will add `IsBlacklistedPair(Atom * beg_atom, Atom *end_atom)`, which will stop `ConnectTheDots()` making bonds to:
* metals
* halogen anions
* noble gasses
But only if they are in separate residue. This will also allow for metalorganic compounds in HETATM residues. I could also inject `HOH` unbinding here.

mwojcikowski commented on Code/GraphMol/FileParsers/PDBParser.cpp:
```diff
@@ -543,8 +560,10 @@ RWMol *PDBBlockToMol(const char *str, bool sanitize, bool removeHs,
-  StandardPDBResidueBondOrders(mol);
+  if (proximityBonding) {
+    ConnectTheDots(mol, ctdIGNORE_H_H_CONTACTS);
+    StandardPDBResidueBondOrders(mol);
```
I was going to say no, but if you had a PDB file in "flavor" that don't encode bonds multiple times (`flavor & 8 ` as defined by Roger) then you could use `StandardPDBResidueBondOrders`. On the other hand I see the `proximityBonding` as a "don't touch bonding at all" flag. This is particularly useful when you've edited the molecule so that it sanitizes and you want to read it back from a "clean" PDB.

mwojcikowski commented:
@greglandrum I have fixed all raised issues. Additionally I've improved bonding removal from metals so that the metalorganic ligands are fine + additional tests. The "inter-residue" bonds are also removed from noble gasses and halogen ions.

mwojcikowski commented:
Also PDBbind v2016 general set stats went up from 13047 to 13082 proteins (my rdkit build includes #1643 since there are proteins larger than 65k atoms).

greglandrum commented on Code/GraphMol/FileParsers/ProximityBonds.cpp:
```diff
@@ -47,6 +47,41 @@ static bool IsBonded(ProximityEntry *p, ProximityEntry *q, unsigned int flags) {
+        return true;
+}
+
+bool IsBlacklistedPair(Atom *beg_atom, Atom *end_atom) {
```
Since this can be called from other places, it should start with a `PRECONDITION` for both `beg_atom` and `end_atom`.

greglandrum commented on Code/GraphMol/FileParsers/ProximityBonds.cpp:
```diff
@@ -47,6 +47,41 @@ static bool IsBonded(ProximityEntry *p, ProximityEntry *q, unsigned int flags) {
+
+bool IsBlacklistedPair(Atom *beg_atom, Atom *end_atom) {
+    AtomPDBResidueInfo *beg_info = (AtomPDBResidueInfo *)beg_atom->getMonomerInfo();
+    AtomPDBResidueInfo *end_info = (AtomPDBResidueInfo *)end_atom->getMonomerInfo();
```
You should also check that neither `beg_info` nor `end_info` is NULL.
I'm not sure that this is a `PRECONDITION`, since there may be a sensible answer if one is NULL?

mwojcikowski replied:
I have made the validation match the one in `StandardPDBDoubleBond()`

greglandrum commented on Code/GraphMol/FileParsers/ProximityBonds.cpp:
```diff
@@ -147,17 +182,53 @@ static void ConnectTheDots_Large(RWMol *mol, unsigned int flags) {
+    Atom *atom = mol->getAtomWithIdx(i);
+    unsigned int elem = atom->getAtomicNum();
+    // detect multivalent Hs, which could happen with ConnectTheDots
+    AtomPDBResidueInfo *atom_info = (AtomPDBResidueInfo *)(atom->getMonomerInfo());
```
You could move this down into the `if(elem == 1...)` block.

greglandrum commented on Code/GraphMol/FileParsers/ProximityBonds.cpp:
```diff
@@ -74,9 +74,7 @@ bool IsBlacklistedPair(Atom *beg_atom, Atom *end_atom) {
 
     AtomPDBResidueInfo *beg_info = (AtomPDBResidueInfo *)beg_atom->getMonomerInfo();
     AtomPDBResidueInfo *end_info = (AtomPDBResidueInfo *)end_atom->getMonomerInfo();
-    if(!beg_info || !beg_info->getMonomerType() != AtomMonomerInfo::PDBRESIDUE)
```
It seems to me like the test for the monomer type being PDBRESIDUE makes sense. Could it be that the problems encountered in the Travis build for this commit happened because it should be: `if(!beg_info || beg_info->getMonomerType() != AtomMonomerInfo::PDBRESIDUE)`?

mwojcikowski replied:
Oh, nice catch. For some reason after failure I assumed that HETATM got `AtomMonomerInfo::OTHER` type, so I ditched that check. It went in back - I did a rebase and merged the last comits.

greglandrum commented:
Thanks @mwojcikowski !
